Compatibility: `FileNotFoundError` with pyfakefs 5.7.3+ and moto when testing AWS S3 upload

**Describe the bug**
I have a test that uses `moto` and `pyfakefs` to test an upload to AWS S3. The test works with pyfakefs 5.7.2 but fails with 5.7.3+. It comes up an following error after several runs.
```
The error is 
FileNotFoundError: [Errno 2] No such file or directory: '/fw/random-file', 
```
The file is created by `moto` and is not a real file.

Comparing the difference between 5.7.2 and 5.7.3, the error is completely gone in every runs when I change the line back to `version_info >= (3,13) ` in https://github.com/pytest-dev/pyfakefs/compare/v5.7.2...v5.7.3#diff-014e42987059a9c5208200707f50ef6a7533a12910844b2032339db797a80f0eL1012

It looks there might have `os` cache compatibility issues for `moto`.

Full logs
```sh
/dev [.venv] % .venv/bin/pytest -s tests --cache-clear -k test_aws_s3_upload
=================================================== test session starts ====================================================
platform darwin -- Python 3.12.9, pytest-8.3.4, pluggy-1.5.0
rootdir: /dev
configfile: pyproject.toml
plugins: pytest_httpserver-1.1.1, pyfakefs-5.7.3, asyncio-0.25.3, timeout-2.3.1
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=function
collected 1 item                                                                                                           
run-last-failure: no previously failed tests, not deselecting items.

tests/test_abc.py check and upload file
check and upload file
.

==================================================== 1 passed in 1.27s =====================================================
/dev [.venv] % .venv/bin/pytest -s tests --cache-clear -k test_aws_s3_upload
=================================================== test session starts ====================================================
platform darwin -- Python 3.12.9, pytest-8.3.4, pluggy-1.5.0
rootdir: /dev
configfile: pyproject.toml
plugins: pytest_httpserver-1.1.1, pyfakefs-5.7.3, asyncio-0.25.3, timeout-2.3.1
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=function
collected 1 item                                                                                                           
run-last-failure: no previously failed tests, not deselecting items.

tests/test_abc.py check and upload file
check and upload file
.

==================================================== 1 passed in 1.23s =====================================================
/dev [.venv] % .venv/bin/pytest -s tests --cache-clear -k test_aws_s3_upload
=================================================== test session starts ====================================================
platform darwin -- Python 3.12.9, pytest-8.3.4, pluggy-1.5.0
rootdir: /dev
configfile: pyproject.toml
plugins: pytest_httpserver-1.1.1, pyfakefs-5.7.3, asyncio-0.25.3, timeout-2.3.1
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=function
collected 1 item                                                                                                           
run-last-failure: no previously failed tests, not deselecting items.

tests/test_abc.py check and upload file
check and upload file
F

========================================================= FAILURES =========================================================
____________________________________________________ test_aws_s3_upload ____________________________________________________
  + Exception Group Traceback (most recent call last):
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/runner.py", line 341, in from_call
  |     result: TResult | None = func()
  |                              ^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/runner.py", line 242, in <lambda>
  |     lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
  |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
  |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 182, in _multicall
  |     return outcome.get_result()
  |            ^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_result.py", line 100, in get_result
  |     raise exc.with_traceback(exc.__traceback__)
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/threadexception.py", line 92, in pytest_runtest_call
  |     yield from thread_exception_runtest_hook()
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
  |     yield
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 95, in pytest_runtest_call
  |     yield from unraisable_exception_runtest_hook()
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
  |     yield
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/logging.py", line 846, in pytest_runtest_call
  |     yield from self._runtest_for(item, "call")
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/logging.py", line 829, in _runtest_for
  |     yield
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/capture.py", line 880, in pytest_runtest_call
  |     return (yield)
  |             ^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 167, in _multicall
  |     teardown.throw(outcome._exception)
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/skipping.py", line 257, in pytest_runtest_call
  |     return (yield)
  |             ^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
  |     res = hook_impl.function(*args)
  |           ^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/runner.py", line 174, in pytest_runtest_call
  |     item.runtest()
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/python.py", line 1627, in runtest
  |     self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_hooks.py", line 513, in __call__
  |     return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_manager.py", line 120, in _hookexec
  |     return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 182, in _multicall
  |     return outcome.get_result()
  |            ^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_result.py", line 100, in get_result
  |     raise exc.with_traceback(exc.__traceback__)
  |   File "/dev/.venv/lib/python3.12/site-packages/pluggy/_callers.py", line 103, in _multicall
  |     res = hook_impl.function(*args)
  |           ^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
  |     result = testfunction(**testargs)
  |              ^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/.venv/lib/python3.12/site-packages/moto/core/models.py", line 122, in wrapper
  |     result = func(*args, **kwargs)
  |              ^^^^^^^^^^^^^^^^^^^^^
  |   File "/dev/tests/test_abc.py", line 35, in test_aws_s3_upload
  |     asyncio.run(__dispatch())
  |   File "/Users/gnought/.pyenv/versions/3.12.9/lib/python3.12/asyncio/runners.py", line 195, in run
  |     return runner.run(main)
  |            ^^^^^^^^^^^^^^^^
  |   File "/Users/gnought/.pyenv/versions/3.12.9/lib/python3.12/asyncio/runners.py", line 118, in run
  |     return self._loop.run_until_complete(task)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/Users/gnought/.pyenv/versions/3.12.9/lib/python3.12/asyncio/base_events.py", line 691, in run_until_complete
  |     return future.result()
  |            ^^^^^^^^^^^^^^^
  |   File "/dev/tests/test_abc.py", line 47, in __dispatch
  |     async with asyncio.TaskGroup() as tg:
  |                ^^^^^^^^^^^^^^^^^^^
  |   File "/Users/gnought/.pyenv/versions/3.12.9/lib/python3.12/asyncio/taskgroups.py", line 71, in __aexit__
  |     return await self._aexit(et, exc)
  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/Users/gnought/.pyenv/versions/3.12.9/lib/python3.12/asyncio/taskgroups.py", line 164, in _aexit
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/Users/gnought/.pyenv/versions/3.12.9/lib/python3.12/asyncio/threads.py", line 25, in to_thread
    |     return await loop.run_in_executor(None, func_call)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/Users/gnought/.pyenv/versions/3.12.9/lib/python3.12/concurrent/futures/thread.py", line 59, in run
    |     result = self.fn(*self.args, **self.kwargs)
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/dev/tests/test_abc.py", line 42, in func
    |     client.upload_file(file, "foobar", Path(file).name)
    |   File "/dev/.venv/lib/python3.12/site-packages/boto3/s3/inject.py", line 145, in upload_file
    |     return transfer.upload_file(
    |            ^^^^^^^^^^^^^^^^^^^^^
    |   File "/dev/.venv/lib/python3.12/site-packages/boto3/s3/transfer.py", line 372, in upload_file
    |     future.result()
    |   File "/dev/.venv/lib/python3.12/site-packages/s3transfer/futures.py", line 103, in result
    |     return self._coordinator.result()
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/dev/.venv/lib/python3.12/site-packages/s3transfer/futures.py", line 264, in result
    |     raise self._exception
    |   File "/dev/.venv/lib/python3.12/site-packages/s3transfer/tasks.py", line 265, in _main
    |     self._submit(transfer_future=transfer_future, **kwargs)
    |   File "/dev/.venv/lib/python3.12/site-packages/s3transfer/upload.py", line 596, in _submit
    |     upload_input_manager.provide_transfer_size(transfer_future)
    |   File "/dev/.venv/lib/python3.12/site-packages/s3transfer/upload.py", line 245, in provide_transfer_size
    |     self._osutil.get_file_size(transfer_future.meta.call_args.fileobj)
    |   File "/dev/.venv/lib/python3.12/site-packages/s3transfer/utils.py", line 262, in get_file_size
    |     return os.path.getsize(filename)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/dev/.venv/lib/python3.12/site-packages/pyfakefs/fake_path.py", line 584, in wrapped
    |     return getattr(os.path, f.__name__)(*args, **kwargs)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "<frozen genericpath>", line 62, in getsize
    | FileNotFoundError: [Errno 2] No such file or directory: '/fw/random-file'
    +------------------------------------
================================================= short test summary info ==================================================
FAILED tests/test_abc.py::test_aws_s3_upload - ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
==================================================== 1 failed in 1.24s =====================================================
```

**How To Reproduce**
I have created a minimal example to reproduce the issue, run
```sh
pytest -s tests --cache-clear -k test_aws_s3_upload
```
The test file
```python
import asyncio
import os
from contextlib import nullcontext as does_not_raise
from pathlib import Path
import botocore.session
import boto3
import botocore
import botocore.client
import botocore.loaders
import pytest
import s3transfer
from moto.core.decorator import mock_aws
from pyfakefs.fake_filesystem import FakeFilesystem


@pytest.fixture(scope="session", autouse=True)
def config(fs_session: FakeFilesystem):

    # ref: https://github.com/getmoto/moto/issues/1682
    # compatible with moto
    fs_session.add_real_paths(
        [str(Path(str(module.__file__)).parent) for module in [boto3, botocore, s3transfer]],
        read_only=True,
        lazy_dir_read=False,
    )
    fs_session.create_file("/fw/zero-file", contents=b"")
    fs_session.create_file("/fw/random-file", contents=os.urandom(1024))

    yield fs_session
    fs_session.clear_cache()

@mock_aws
def test_aws_s3_upload():
    with does_not_raise():
        asyncio.run(__dispatch())

async def __dispatch():
    def func(file):
        with does_not_raise():
            print("check and upload file")
            assert Path(file).exists()
            client.upload_file(file, "foobar", Path(file).name)

    client = boto3.client("s3")
    client.create_bucket(Bucket="foobar")
    tasks = []
    async with asyncio.TaskGroup() as tg:
        tasks = [tg.create_task(asyncio.to_thread(func, file)) for file in ["/fw/zero-file", "/fw/random-file"]]
    return [task.result() for task in tasks]
```

**Your environment**
macOS-15.3.1-arm64-arm-64bit
Python 3.12.9 (main, Feb 24 2025, 17:54:20) [Clang 16.0.0 (clang-1600.0.26.6)]
pyfakefs 5.7.3
pytest 8.3.4
moto 5.1.0
boto3 1.36.26

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -20,6 +20,8 @@ The released versions correspond to PyPI releases.
 ### Fixes
 * fixed a problem with flushing if writing over the buffer end
   (see [#1120](../../issues/1120))
+* fixed a regression that could break tests under Posix in Python 3.12
+  (see [#1126](../../issues/1126))
 
 ## [Version 5.7.4](https://pypi.python.org/pypi/pyfakefs/5.7.4) (2025-01-14)
 Minor bugfix release.

--- a/pyfakefs/fake_filesystem_unittest.py
+++ b/pyfakefs/fake_filesystem_unittest.py
@@ -130,6 +130,54 @@ class TempfilePatcher:
         tempfile.tempdir = None
 
 
+class LineCachePatcher:
+    """Handles linecache patching for newer Python versions."""
+
+    def __init__(self):
+        self.linecache_updatecache = None
+        self.linecache_checkcache = None
+
+    def start_patching(self):
+        if self.linecache_updatecache is not None:
+            return
+
+        def checkcache(filename=None):
+            """Calls the original linecache.checkcache making sure no fake OS calls
+            are used."""
+            with use_original_os():
+                return self.linecache_checkcache(filename)
+
+        def updatecache(filename, module_globals=None):
+            """Calls the original linecache.updatecache making sure no fake OS calls
+            are used."""
+
+            with use_original_os():
+                # workaround for updatecache problem with pytest under Windows, see #1096
+                if not filename.endswith(r"pytest.exe\__main__.py"):
+                    return self.linecache_updatecache(filename, module_globals)
+                return []
+
+        if (
+            sys.version_info >= (3, 13)
+            or sys.platform == "win32"
+            and sys.version_info >= (3, 12)
+        ):
+            # in linecache, 'os' is now imported locally, which involves the
+            # dynamic patcher, therefore we patch the affected functions
+            self.linecache_updatecache = linecache.updatecache
+            linecache.updatecache = updatecache
+            self.linecache_checkcache = linecache.checkcache
+            linecache.checkcache = checkcache
+
+    def stop_patching(self):
+        if self.linecache_updatecache is None:
+            return
+        linecache.updatecache = self.linecache_updatecache
+        linecache.checkcache = self.linecache_checkcache
+        self.linecache_updatecache = None
+        self.linecache_checkcache = None
+
+
 def patchfs(
     _func: Optional[Callable] = None,
     *,
@@ -616,9 +664,8 @@ class Patcher:
         # save the original open function for use in pytest plugin
         self.original_open = open
         self.patch_open_code = patch_open_code
-        self.linecache_updatecache = None
-        self.linecache_checkcache = None
         self.tempfile_patcher = TempfilePatcher()
+        self.linecache_patcher = LineCachePatcher()
 
         if additional_skip_names is not None:
             skip_names = [
@@ -691,21 +738,6 @@ class Patcher:
         self._paused = False
         self.has_copy_file_range = False
         self.has_copy_file = False
-
-    def checkcache(self, filename=None):
-        """Calls the original linecache.checkcache making sure no fake OS calls
-        are used."""
-        with use_original_os():
-            return self.linecache_checkcache(filename)
-
-    def updatecache(self, filename, module_globals=None):
-        """Calls the original linecache.updatecache making sure no fake OS calls
-        are used."""
-        with use_original_os():
-            # workaround for updatecache problem with pytest under Windows, see #1096
-            if not filename.endswith(r"pytest.exe\__main__.py"):
-                return self.linecache_updatecache(filename, module_globals)
-            return []
 
     @classmethod
     def clear_fs_cache(cls) -> None:
@@ -1028,14 +1060,7 @@ class Patcher:
             self._patching = True
             self._paused = False
 
-            if sys.version_info >= (3, 12):
-                # in linecache, 'os' is now imported locally, which involves the
-                # dynamic patcher, therefore we patch the affected functions
-                self.linecache_updatecache = linecache.updatecache
-                linecache.updatecache = self.updatecache
-                self.linecache_checkcache = linecache.checkcache
-                linecache.checkcache = self.checkcache
-
+            self.linecache_patcher.start_patching()
             self.tempfile_patcher.start_patching()
 
             self.patch_modules()
@@ -1156,9 +1181,7 @@ class Patcher:
                 self._dyn_patcher.cleanup()
                 sys.meta_path.pop(0)
             self.tempfile_patcher.stop_patching()
-            if self.linecache_updatecache is not None:
-                linecache.updatecache = self.linecache_updatecache
-                linecache.checkcache = self.linecache_checkcache
+            self.linecache_patcher.stop_patching()
             self._set_glob_os_functions()
 
     @property

